spec/v2/providers/eventarc.spec.ts

```diff
@@ -0,0 +1,161 @@
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+import { expect } from 'chai';
+import * as options from '../../../src/v2/options';
+import * as eventarc from '../../../src/v2/providers/eventarc';
+import { FULL_OPTIONS } from './helpers';
+
+const ENDPOINT_EVENT_TRIGGER = {
+  eventType: 'event-type',
+  retry: false,
+};
+
+describe('v2/eventarc', () => {
+  describe('onCustomEventPublished', () => {
+    beforeEach(() => {
+      process.env.GCLOUD_PROJECT = 'aProject';
+    });
+
+    afterEach(() => {
+      options.setGlobalOptions({});
+      delete process.env.GCLOUD_PROJECT;
+    });
+
+    it('should create a minimal trigger/endpoint with bucket', () => {
+      const result = eventarc.onCustomEventPublished('event-type', () => 42);
+
+      expect(result.__endpoint).to.deep.equal({
+        platform: 'gcfv2',
+        labels: {},
+        eventTrigger: {
+          ...ENDPOINT_EVENT_TRIGGER,
+          channel: 'locations/us-central1/channels/firebase',
+        },
+      });
+    });
+
+    it('should create a minimal trigger/endpoint with opts', () => {
+      const result = eventarc.onCustomEventPublished(
+        { eventType: 'event-type', region: 'us-west1' },
+        () => 42
+      );
+
+      expect(result.__endpoint).to.deep.equal({
+        platform: 'gcfv2',
+        labels: {},
+        eventTrigger: {
+          ...ENDPOINT_EVENT_TRIGGER,
+          channel: 'locations/us-central1/channels/firebase',
+        },
+        region: ['us-west1'],
+      });
+    });
+
+    it('should create a minimal trigger with bucket with opts', () => {
+      const result = eventarc.onCustomEventPublished(
+        {
+          eventType: 'event-type',
+          channel: 'locations/us-west1/channels/my-channel',
+          filters: { foo: 'bar' },
+        },
+        () => 42
+      );
+
+      expect(result.__endpoint).to.deep.equal({
+        platform: 'gcfv2',
+        labels: {},
+        eventTrigger: {
+          ...ENDPOINT_EVENT_TRIGGER,
+          channel: 'locations/us-west1/channels/my-channel',
+          eventFilters: {
+            foo: 'bar',
+          },
+        },
+      });
+    });
+
+    it('should create a complex trigger/endpoint with appropriate values', () => {
+      const result = eventarc.onCustomEventPublished(
+        {
+          ...FULL_OPTIONS,
+          eventType: 'event-type',
+          channel: 'locations/us-west1/channels/my-channel',
```

Comment:
On the functions SDK side I'm only doing the: if channel is not specified, default it to 'locations/us-central1/channels/firebase'. The rest of the partial channel resource name resolution (and validation) is done on the CLI side.

The reason it's done this way is because if channel is not specified, in the container contract I can't distinguish between: `it's not specified and expected to be resolved to default` vs `not specified and should remain undefined`.

Reply:
Sounds good. And the more code we put in the "agent" (e.g. CLI, control plane) the less we have to redefine in the SDKs.